### PR TITLE
[vcpkg] Make CMakelists nicer 😁

### PR DESCRIPTION
**Pull request by strega-nil**
Do a few things:

* Add the "VCPKG_DEVELOPMENT_WARNINGS" flag
  * setting "WERROR" will also set this flag
  * This flag is set by default
  * on GCC, this will pass '-Wall -Wextra -Wpedantic -Werror'
  * on MSVC, this will pass '-W4 -WX'
* On Visual Studio 2017 and later, pass '-permissive-'
* Support Visual Studio 2015 in the CMakeLists.txt
* Fix the fallout of these changes on the source code

### CODE REVIEW

**Comment by ubsan** [comment]
Depends on #7637

**Comment by ubsan** [comment]
(Please don't merge! I want to write a better commit message)

File: toolsrc/src/vcpkg/base/system.cpp [hunk 1]
```diff
@@ -381,6 +382,8 @@ namespace vcpkg
             "CreateProcessW() returned ", exit_code, " after ", static_cast<int>(timer.microseconds()), " us\n");
         return static_cast<int>(exit_code);
 #else
+        // TODO(ubsan): fix this to actually have a clean environment
```
Comment (Rastaban) [inline toolsrc/src/vcpkg/base/system.cpp]: This is currently a known issue that we haven't addressed yet.  If you are signing up to be the one who will definitely fix this issue then I am okay with the comment.  Otherwise I would rather you put a short explanation that would be useful to someone else who may notice it and address the issue.

File: toolsrc/src/vcpkg/base/system.cpp [hunk 2]
```diff
@@ -549,10 +552,7 @@ namespace vcpkg
         return Strings::to_utf8(ret);
     }
 #else
-    Optional<std::string> System::get_registry_string(void* base_hkey, StringView sub_key, StringView valuename)
-    {
-        return nullopt;
-    }
+    Optional<std::string> System::get_registry_string(void*, StringView, StringView) { return nullopt; }
```
Comment (Rastaban) [inline toolsrc/src/vcpkg/base/system.cpp]: Did clang-format put this all on one line?

File: toolsrc/src/vcpkg/base/system.print.cpp [hunk 3]
```diff
@@ -21,6 +22,8 @@ namespace vcpkg::System
             System::print2(message);
             SetConsoleTextAttribute(console_handle, original_color);
 #else
+            // TODO(ubsan): add color handling code
```
Comment (Rastaban) [inline toolsrc/src/vcpkg/base/system.print.cpp]: Same thing as earlier about putting `// TODO(ubsan)`

File: toolsrc/src/vcpkg/export.cpp [hunk 4]
```diff
@@ -360,6 +360,12 @@ namespace vcpkg::Export
                 }
             };
 
+#if _MSC_VER <= 1900
```
Comment (ubsan) [inline toolsrc/src/vcpkg/export.cpp]: check if `_MSC_VER` is defined

File: toolsrc/src/vcpkg/base/system.cpp (hunk 2 above)
Comment (vicroms) [reply toolsrc/src/vcpkg/base/system.cpp]: Our `.clang-format` has a limit of 120 columns so I think it did convert this into a one-liner.

File: toolsrc/src/vcpkg/build.cpp [hunk 5]
```diff
@@ -216,7 +216,7 @@ namespace vcpkg::Build
                                   }));
     }
 
-    std::unordered_map<std::string, std::string> make_env_passthrough(const PreBuildInfo& pre_build_info)
+    static auto make_env_passthrough(const PreBuildInfo& pre_build_info) -> std::unordered_map<std::string, std::string>
```
Comment (cbezault) [inline toolsrc/src/vcpkg/build.cpp]: Why the change in return type syntax?

Comment (ubsan) [reply toolsrc/src/vcpkg/build.cpp]: with the static and the type name, it was unclear what the name of the function is, and I thought "as long as I'm changing it"

Comment (cbezault) [reply toolsrc/src/vcpkg/build.cpp]: fair

File: toolsrc/CMakeLists.txt [hunk 6]
```diff
@@ -73,31 +57,70 @@ if(CLANG)
...
+    add_test(NAME default COMMAND vcpkg-test)
 
     if (VCPKG_BUILD_BENCHMARKING)
         target_compile_options(vcpkg-test PRIVATE -DCATCH_CONFIG_ENABLE_BENCHMARKING)
     endif()
+
+    find_program(CLANG_FORMAT clang-format)
+    if (CLANG_FORMAT)
+        file(GLOB_RECURSE VCPKG_FORMAT_SOURCES CONFIGURE_DEPENDS src/*.cpp include/pch.h include/vcpkg/*.h include/vcpkg-test/*.h)
+        add_custom_target(format COMMAND ${CLANG_FORMAT} -i -verbose ${VCPKG_FORMAT_SOURCES})
```
Comment (grdowns) [inline toolsrc/CMakeLists.txt]: We may need to surround `${CLANG_FORMAT}` in quotes in case the path contains spaces

File: toolsrc/src/vcpkg/tools.cpp [hunk 7]
```diff
@@ -274,6 +274,9 @@ namespace vcpkg
             const auto& program_files_32_bit = System::get_program_files_32_bit();
             if (const auto pf = program_files_32_bit.get())
                 out_candidate_paths.push_back(*pf / "CMake" / "bin" / "cmake.exe");
+#else
+            // TODO(ubsan): figure out if this should do anything on non-Windows
```
Comment (grdowns) [inline toolsrc/src/vcpkg/tools.cpp]: Same feedback here that @Rastaban pointed out about adding this as your own responsibility

File: toolsrc/src/vcpkg/tools.cpp [hunk 8]
```diff
@@ -362,6 +365,9 @@ Type 'NuGet help <command>' for help on a specific command.
             const auto& program_files_32_bit = System::get_program_files_32_bit();
             if (const auto pf = program_files_32_bit.get())
                 out_candidate_paths.push_back(*pf / "git" / "cmd" / "git.exe");
+#else
+            // TODO(ubsan): figure out if this should do anything on non-windows
```
Comment (grdowns) [inline toolsrc/src/vcpkg/tools.cpp]: Here as well

File: toolsrc/CMakeLists.txt (hunk 6 above)
Comment (grdowns) [inline toolsrc/CMakeLists.txt]: We may need to surround `${CLANG_FORMAT}` in quotes in case the path contains spaces

Comment (ubsan) [reply toolsrc/CMakeLists.txt]: I don't think this is true
